File: c/11.c
#include "common.h"

#define GRID_SIZE 10
int orig[GRID_SIZE][GRID_SIZE];
int grid[GRID_SIZE][GRID_SIZE];

void parse() {
  FILE *f = fopen(INFILE("11.txt"), "r");
  char buf[2048];

  int row = 0;
  while (GETLINE(buf, f)) {
    for (int col = 0; col < GRID_SIZE; col++) {
      orig[row][col] = buf[col] - '0';
    }
    row++;
  }

  fclose(f);
}

void dump() {
  for (int row = 0; row < GRID_SIZE; row++) {
    for (int col = 0; col < GRID_SIZE; col++)
      printf("%d", grid[row][col]);
    printf("\n");
  }
}

int step() {
  for (int row = 0; row < GRID_SIZE; row++)
    for (int col = 0; col < GRID_SIZE; col++)
      grid[row][col] += 1;
  int flashed = 0;
  bool changed = true;
  while (changed) {
    changed = false;
    for (int row = 0; row < GRID_SIZE; row++)
      for (int col = 0; col < GRID_SIZE; col++) {
        if (grid[row][col] <= 9) continue;
        grid[row][col] = 0;
        changed = true;
        flashed++;
        for (int r = row - 1; r <= row + 1; r++)
          for (int c = col - 1; c <= col + 1; c++)
            if (0 <= r && r < GRID_SIZE && 
                0 <= c && c < GRID_SIZE &&
                grid[r][c] > 0) {
              grid[r][c]++;
            }
      }
  }
  return flashed;
}

void part1() {
  memcpy(grid, orig, sizeof(grid));
  u64 result = 0;
  for (int i = 0; i < 100; i++) {
    result += step();
  }
  printf("Part 1: %llu\n", result);
}

void part2() {
  memcpy(grid, orig, sizeof(grid));
  int i;
  for (i = 1; step() != GRID_SIZE * GRID_SIZE; i++)
    ;
  printf("Part 2: %i\n", i);
}


int main() {
  parse();
  part1();
  part2();
  return 0;
}

File: python/11.py
from common import *

with open(infile('11.txt')) as f:
    data = [[int(x) for x in line.strip()] for line in f]
    data = grid_to_dict(data)

def dump(grid):
    for j in range(10):
        for i in range(10):
            print(grid[(i, j)], end="")
        print()

def takeStep(grid):
    for k in grid:
        grid[k] += 1

    flashed = 0
    changed = True
    while changed:
        changed = False
        for k, val in grid.items():
            if val > 9:
                grid[k] = 0
                flashed += 1
                changed = True
                for n in neighbours(*k, diag=True):
                    if n in grid and grid[n] > 0:
                        grid[n] += 1
    return flashed


def part1():
    grid = data.copy()
    count = 0
    for _ in range(100):
        count += takeStep(grid)
    print("Part 1:", count)

def part2():
    grid = data.copy()
    count = 1
    res = takeStep(grid)
    while res != 100:
        res = takeStep(grid)
        count += 1
    print("Part 2:", count)


def animate():
    from PIL import Image
    from random import randint

    grid = {}
    SIZE = 150
    for j in range(SIZE):
        for i in range(SIZE):
            grid[(i, j)] = randint(1, 9) 

    FRAMES = 120
    for i in tqdm(range(FRAMES)):
        im = Image.new('L', (SIZE, SIZE))
        im.putgrid([grid[(i, j)]*25 for j in range(SIZE) for i in range(SIZE)])
        im = im.resize((500, 500), Image.NEAREST)
        im.save("frames/frame%03d.png" % i)
        takeStep(grid)

part1()
part2()
# animate()